Alluvial/Stream.cs

```diff
@@ -132,5 +133,40 @@ public static ICursor CreateCursor<TData>(this IStream<TData> stream)
 
             return projection;
         }
+
+        public static IStream<TData> Trace<TData>(
+            this IStream<TData> stream,
+            Action<IStreamQuery> onSendQuery = null,
+            Action<IStreamQuery, IStreamBatch<TData>> onResults = null)
+        {
+            onSendQuery = onSendQuery ??
+                          (q => System.Diagnostics.Trace.WriteLine(
+                              string.Format("Query: stream {0} @ cursor position {1}",
+                                            stream.Id,
+                                            (object) q.Cursor.Position)));
+
+            onResults = onResults ?? 
+                ((q, streamBatch) => System.Diagnostics.Trace.WriteLine(
+                    string.Format("Fetched: stream {0} batch of {1}, now @ cursor position {2}",
+                                  stream.Id,
+                                  streamBatch.Count,
+                                  (object) q.Cursor.Position)));
+
+            return Create<TData>(
+                id: stream.Id,
+                query: async q =>
+                {
+                    onSendQuery(q);
+
+                    var streamBatch = await stream.Fetch(q);
+
+                    onResults(q, streamBatch);
+
+                    return streamBatch;
+                },
+                advanceCursor: (q, b) =>

```

Comment:
The default implementation if you don't pass something will advance the cursor, resulting in a double advance. I'd like to figure it a way to not have that happen although there are valid used for it, e.g. Requery. 



> On Jan 31, 2015, at 2:49 PM, Nathan Black <notifications@github.com> wrote:
> 
> In Alluvial/Stream.cs:
> 
> > +                                  streamBatch.Count,
> > +                                  (object) q.Cursor.Position)));
> > +
> > +            return Create<TData>(
> > +                id: stream.Id,
> > +                query: async q =>
> > +                {
> > +                    onSendQuery(q);
> > +
> > +                    var streamBatch = await stream.Fetch(q);
> > +
> > +                    onResults(q, streamBatch);
> > +
> > +                    return streamBatch;
> > +                },
> > +                advanceCursor: (q, b) =>
> Why is a noop cursor specified instead of omitting this (it's optional, right?)
> 
> —
> Reply to this email directly or view it on GitHub.
>